base/array.jl

```diff
@@ -3133,14 +3133,17 @@ setdiff!(  v::AbstractVector, itrs...) = _shrink!(setdiff!, v, itrs)
 
 vectorfilter(T::Type, f, v) = T[x for x in v if f(x)]
 
-function _shrink(shrinker!::F, itr, itrs) where F
+function intersect(itr, itrs...)
     T = promote_eltype(itr, itrs...)
-    keep = shrinker!(Set{T}(itr), itrs...)
+    keep = intersect!(Set{T}(itr), itrs...)
     vectorfilter(T, _shrink_filter!(keep), itr)
 end
 
-intersect(itr, itrs...) = _shrink(intersect!, itr, itrs)
-setdiff(  itr, itrs...) = _shrink(setdiff!, itr, itrs)
+function setdiff(itr, itrs...)
+    T = promote_eltype(itr, itrs...)
+    keep = setdiff!(Set{T}(itr), itrs...)
+    vectorfilter(eltype(itr), _shrink_filter!(keep), itr)
```

Comment:
Would this work? Possibly I'm misunderstanding something, but I don't see why is promotion necessary, given that `setdiff!` never adds elements to the set.

```suggestion
    T = eltype(itr)
    keep = setdiff!(Set{T}(itr), itrs...)
    vectorfilter(T, _shrink_filter!(keep), itr)
```

Reply:
let's see